linkerd/proxy/transport/src/prefix.rs

```diff
@@ -0,0 +1,75 @@
+use crate::io;
+use futures::prelude::*;
+use linkerd2_error::Error;
+use linkerd2_stack::layer;
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+    time::Duration,
+};
+use tokio::time;
+use tower::util::ServiceExt;
+use tracing::debug;
+
+#[derive(Copy, Clone)]
+pub struct Prefix<S> {
+    inner: S,
+    capacity: usize,
+    timeout: Duration,
+}
+
+#[derive(Debug)]
+pub struct ReadTimeout(());
+
+impl<S> Prefix<S> {
+    pub fn new(inner: S, capacity: usize, timeout: Duration) -> Self {
+        Self {
+            inner,
+            capacity,
+            timeout,
+        }
+    }
+
+    pub fn layer(
+        capacity: usize,
+        timeout: Duration,
+    ) -> impl layer::Layer<S, Service = Prefix<S>> + Clone {
+        layer::mk(move |inner| Self::new(inner, capacity, timeout))
+    }
+}
+
+impl<S, I> tower::Service<I> for Prefix<S>
+where
+    I: io::Peekable + Send + 'static,
+    S: tower::Service<io::PrefixedIo<I>, Response = ()> + Clone + Send + 'static,
+    S::Error: Into<Error>,
+    S::Future: Send,
+{
+    type Response = ();
+    type Error = Error;
+    type Future = Pin<Box<dyn Future<Output = Result<(), Error>> + Send + 'static>>;
+
+    fn poll_ready(&mut self, _: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(().into()))
+    }
+
+    fn call(&mut self, io: I) -> Self::Future {
+        debug!(capacity = %self.capacity, "Buffering prefix");
```

Comment:
nit: `usize` is a `tracing` primitive `Value` type, so we don't really need to use `fmt::Display` here:
```suggestion
        debug!(capacity = self.capacity, "Buffering prefix");
```